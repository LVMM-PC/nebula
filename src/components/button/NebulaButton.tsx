import { Component, Prop, Vue, Watch } from "vue-property-decorator";
import Wave from "../_util/wave";
import Icon from "../icon";
import buttonTypes from "./buttonTypes";
import { mixins } from "vue-class-component";

const rxTwoCNChar = /^[\u4e00-\u9fa5]{2}$/;
const isTwoCNChar = rxTwoCNChar.test.bind(rxTwoCNChar);

const props = buttonTypes();

export type ButtonType = "default" | "primary" | "ghost" | "dashed" | "danger";
export type ButtonShape = "circle" | "circle-outline";
export type ButtonSize = "small" | "default" | "large";
export type ButtonHTMLType = "submit" | "button" | "reset";

type loadingType = boolean | { delay?: number };

@Component({})
export class BaseButtonProps extends Vue {
  @Prop({ default: "nebula-btn", type: String })
  prefixCls?: string;
  @Prop({ type: String })
  type?: ButtonType;
  @Prop({ default: "button", type: String })
  htmlType?: ButtonHTMLType;
  @Prop({ type: String })
  icon?: string;
  @Prop({ type: String })
  shape?: ButtonShape;
  @Prop({ type: String })
  size?: ButtonSize;
  @Prop({ type: [Boolean, Object] })
  loading?: loadingType;
  @Prop({ type: Boolean })
  disabled?: boolean;
  @Prop({ type: Boolean })
  ghost?: boolean;
  @Prop({ type: Boolean })
  block?: boolean;

  constructor(props) {
    super(props);
  }
}

@Component({
  // @ts-ignore
  __NEBULA_BUTTON: true
})
export default class NebulaButton extends mixins(BaseButtonProps) {
  static __NEBULA_BUTTON: any = true;
  public delayTimeout?: any;
  public sizeMap?: any = {
    large: "lg",
    small: "sm"
  };
  public sLoading?: any = !!this.loading;
  public hasTwoCNChar?: any = false;

  constructor(props) {
    super(props);
  }

  get classes() {
    const {
      prefixCls,
      type,
      shape,
      size,
      hasTwoCNChar,
      sLoading,
      ghost,
      block,
      sizeMap
    } = this;
    const sizeCls = sizeMap[size] || "";
    return {
      [`${prefixCls}`]: true,
      [`${prefixCls}-${type}`]: type,
      [`${prefixCls}-${shape}`]: shape,
      [`${prefixCls}-${sizeCls}`]: sizeCls,
      [`${prefixCls}-loading`]: sLoading,
      [`${prefixCls}-background-ghost`]: ghost || type === "ghost",
      [`${prefixCls}-two-chinese-chars`]: hasTwoCNChar,
      [`${prefixCls}-block`]: block
    };
  }

  @Watch("loading")
  onLoadingChanged(val: loadingType) {
    clearTimeout(this.delayTimeout);
    if (typeof val !== "boolean" && val && val.delay) {
      this.delayTimeout = setTimeout(() => {
        this.sLoading = !!val;
      }, val.delay);
    } else {
      this.sLoading = !!val;
    }
  }

  initData() {}

  created() {
    this.initData();
  }

  mounted() {
    this.fixTwoCNChar();
  }

  updated() {
    this.fixTwoCNChar();
  }

  beforeDestroy() {
    // if (this.timeout) {
    //   clearTimeout(this.timeout)
    // }
    if (this.delayTimeout) {
      clearTimeout(this.delayTimeout);
    }
  }

  fixTwoCNChar() {
    // Fix for HOC usage like <FormatMessage />
    const node = this.$el;
    const buttonText = node.textContent || node.innerText;
    if (this.isNeedInserted() && isTwoCNChar(buttonText)) {
      if (!this.hasTwoCNChar) {
        this.hasTwoCNChar = true;
      }
    } else if (this.hasTwoCNChar) {
      this.hasTwoCNChar = false;
    }
  }

  handleClick(event) {
    // this.clicked = true
    // clearTimeout(this.timeout)
    // this.timeout = setTimeout(() => (this.clicked = false), 500)
    this.$emit("click", event);
  }

  insertSpace(child, needInserted) {
    const SPACE = needInserted ? " " : "";
    if (typeof child.text === "string") {
      let text = child.text.trim();
      if (isTwoCNChar(text)) {
        text = text.split("").join(SPACE);
      }
      return <span>{text}</span>;
    }
    return child;
  }

  isNeedInserted() {
    const { icon, $slots } = this;
    return $slots.default && $slots.default.length === 1 && !icon;
  }

  render() {
    const {
      htmlType,
      classes,
      icon,
      disabled,
      handleClick,
      sLoading,
      $slots,
      $attrs,
      $listeners
    } = this;
    const buttonProps = {
      props: {},
      attrs: {
        ...$attrs,
        type: htmlType,
        disabled
      },
      class: classes,
      on: {
        ...$listeners,
        click: handleClick
      }
    };
    const iconType = sLoading ? "loading" : icon;
    const iconNode = iconType ? <Icon type={iconType} /> : null;
    const kids =
      $slots.default && $slots.default.length === 1
        ? this.insertSpace($slots.default[0], this.isNeedInserted())
        : $slots.default;
    if ("href" in $attrs) {
      return (
        <a {...buttonProps}>
          {iconNode}
          {kids}
        </a>
      );
    } else {
      return (
        <Wave>
          <button {...buttonProps}>
            {iconNode}
            {kids}
          </button>
        </Wave>
      );
    }
  }
}
