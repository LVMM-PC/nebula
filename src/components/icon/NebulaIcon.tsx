import { Component, Prop, Vue } from "vue-property-decorator";

@Component({
  name: "NebulaIcon"
})
export default class NebulaIcon extends Vue {
  @Prop({ default: null, type: String })
  private type?: string;

  @Prop({ default: null, type: String })
  private title?: string;

  @Prop({ default: false, type: Boolean })
  private spin?: boolean;

  @Prop()
  private timeout?: any;

  public prefixCls: string = "nebula-icon";

  public clicked: boolean = false;

  get classes() {
    let prefixCls = this.prefixCls;
    let type = this.type;
    let spin = this.spin || (type && type.indexOf("loading") === 0);
    return [
      prefixCls,
      {
        [`${prefixCls}-${type}`]: type,
        [`${prefixCls}-spin`]: spin
      }
    ];
  }

  handleClick(event) {
    if (this.clicked) {
      return;
    }

    this.clicked = true;
    clearTimeout(this.timeout);
    this.timeout = setTimeout(() => (this.clicked = false), 500);
    this.$emit("click", event);
  }

  render() {
    const { title, classes, handleClick, $listeners } = this;
    const iconProps = {
      attrs: {
        title
      },
      class: classes,
      on: {
        ...$listeners,
        click: handleClick
      }
    };
    return <i {...iconProps} />;
  }
}
