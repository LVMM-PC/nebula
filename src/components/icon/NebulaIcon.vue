<template>
  <span
    v-bind="$attrs"
    :class="classes"/>
</template>

<script lang="ts">
import { Component, Prop, Vue } from "vue-property-decorator";

@Component
export default class NebulaIcon extends Vue {
  @Prop({ default: null, type: String })
  private type?: string;

  @Prop({ default: false, type: Boolean })
  private spin?: boolean;

  public prefixCls: string = "nebula-icon";

  get classes() {
    let prefixCls = this.prefixCls;
    let type = this.type;
    let spin = this.spin || (type && type.indexOf("loading") === 0);
    return [
      prefixCls,
      {
        [`${prefixCls}-${type}`]: type,
        [`${prefixCls}-spin`]: spin
      }
    ];
  }
}
</script>
