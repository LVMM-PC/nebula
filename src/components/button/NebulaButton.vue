<script lang="tsx">
import { Component, Prop, Vue, Watch } from "vue-property-decorator";
import { VNode } from "vue";
import NebulaIcon from "../icon/NebulaIcon.vue";

const rxTwoCNChar = /^[\u4e00-\u9fa5]{2}$/;
const isTwoCNChar = (str: any) => {
  if (typeof str !== "string") {
    return;
  }
  return rxTwoCNChar.test(str);
};
const isString = (str: any) => {
  return typeof str === "string";
};

export type ButtonType = "default" | "primary" | "ghost" | "dashed" | "danger";
export type ButtonShape = "circle" | "circle-outline";
export type ButtonSize = "small" | "default" | "large";
export type ButtonHTMLType = "submit" | "button" | "reset";

@Component({
  components: {
    NebulaIcon
  }
})
export default class NebulaButton extends Vue {
  static __NEBULA_BUTTON: boolean = true;

  @Prop({ default: null, type: String })
  private icon?: string;

  @Prop({ default: null, type: String })
  private shape?: ButtonShape;

  @Prop({ default: null, type: String })
  private type?: ButtonType;

  @Prop({ default: null, type: String })
  private size?: ButtonSize;

  @Prop({ default: null, type: Boolean })
  private ghost?: boolean;

  @Prop({ default: null, type: Boolean })
  private disabled?: boolean;

  @Prop({ default: null, type: [Boolean, Object] })
  private loading?: boolean | { delay: number };

  @Prop({ default: null, type: Boolean })
  private block?: boolean;

  @Prop({ default: "button", type: String })
  private htmlType?: ButtonHTMLType;

  @Watch("loading")
  onLoadingChanged(val: boolean | { delay?: number }) {
    const currentLoading = this.loading;

    if (currentLoading) {
      clearTimeout(this.delayTimeout);
    }

    if (typeof val === "object" && val && val.delay) {
      this.delayTimeout = setTimeout(() => {
        this.stateLoading = true;
      }, val.delay);
    } else if (typeof val === "boolean") {
      this.stateLoading = val;
    }
  }

  get sizeCls() {
    // large => lg
    // small => sm
    let sizeCls = "";
    switch (this.size) {
      case "large":
        sizeCls = "lg";
        break;
      case "small":
        sizeCls = "sm";
        break;
      default:
        break;
    }
    return sizeCls;
  }

  get component() {
    if (this.$attrs["href"]) {
      return "a";
    } else {
      return "button";
    }
  }

  get iconType() {
    if (this.stateLoading) {
      return "loading";
    } else {
      return this.icon;
    }
  }

  mounted() {
    this.fixTwoCNChar();
  }

  updated() {
    this.fixTwoCNChar();
  }

  beforeDestroy() {
    if (this.delayTimeout) {
      clearTimeout(this.delayTimeout);
    }
  }

  public prefixCls: string = "nebula-btn";

  public hasTwoCNChar: boolean = false;

  public stateLoading: boolean = !!this.loading;

  public delayTimeout: any;

  public classes() {
    let block = this.block;
    let children = this.children();
    let ghost = this.ghost;
    let hasTwoCNChar = this.hasTwoCNChar;
    let icon = this.icon;
    let loading = this.stateLoading;
    let prefixCls = this.prefixCls;
    let shape = this.shape;
    let sizeCls = this.sizeCls;
    let type = this.type;
    return [
      prefixCls,
      {
        [`${prefixCls}-${shape}`]: shape,
        [`${prefixCls}-${sizeCls}`]: sizeCls,
        [`${prefixCls}-${type}`]: type,
        [`${prefixCls}-background-ghost`]: ghost,
        [`${prefixCls}-block`]: block,
        [`${prefixCls}-icon-only`]: !children && icon,
        [`${prefixCls}-loading`]: loading,
        [`${prefixCls}-two-chinese-chars`]: hasTwoCNChar
      }
    ];
  }

  public insertSpace(child: VNode, needInserted: boolean) {
    if (child == null) {
      return;
    }
    const SPACE = needInserted ? " " : "";
    let childText =
      child.children && child.children[0] && child.children[0].text;
    if (
      typeof child !== "string" &&
      typeof child !== "number" &&
      isString(child.tag) &&
      child.children &&
      child.children.length === 1 &&
      childText &&
      isString(childText) &&
      isTwoCNChar(childText)
    ) {
      return childText.split("").join(SPACE);
    }
    let text = child.text;
    if (typeof text === "string") {
      if (isTwoCNChar(text)) {
        text = text.split("").join(SPACE);
      }
      return <span>{text}</span>;
    }
    return child;
  }

  public handleClick(event: Event) {
    if (this.ghost) {
      event.preventDefault();
      event.stopPropagation();
      event.stopImmediatePropagation();
    } else {
      this.$emit("click", event);
    }
  }

  public isNeedInserted() {
    const { iconType, $slots } = this;
    return $slots.default && $slots.default.length === 1 && !iconType;
  }

  public fixTwoCNChar() {
    const node = this.$el as HTMLElement;
    const buttonText = node.textContent || node.innerText;
    if (this.isNeedInserted() && isTwoCNChar(buttonText)) {
      if (!this.hasTwoCNChar) {
        this.hasTwoCNChar = true;
      }
    } else if (this.hasTwoCNChar) {
      this.hasTwoCNChar = false;
    }
  }

  public children() {
    return this.$slots.default;
  }

  render() {
    const {
      $attrs,
      $listeners,
      disabled,
      handleClick,
      htmlType,
      iconType
    } = this;
    let classes = this.classes();
    const buttonProps = {
      props: {},
      attrs: {
        ...$attrs,
        type: htmlType,
        disabled
      },
      class: classes,
      on: {
        ...$listeners,
        click: handleClick
      }
    };

    const iconProps = {
      props: {
        type: iconType
      }
    };
    const iconNode = iconType ? <NebulaIcon {...iconProps} /> : "";
    let children = this.children();
    const kids = children
      ? children.map(child => {
          return this.insertSpace(child, this.isNeedInserted());
        })
      : null;

    switch (this.component) {
      case "a":
        return (
          <a {...buttonProps}>
            {iconNode}
            {kids}
          </a>
        );
      case "button":
        return (
          <button {...buttonProps}>
            {iconNode}
            {kids}
          </button>
        );
    }
  }
}
</script>
