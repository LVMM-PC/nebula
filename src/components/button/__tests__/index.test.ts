import { shallowMount, mount } from "@vue/test-utils";

// import {NebulaButton,NebulaIcon} from "../../../"
import NebulaButton from "../NebulaButton.vue";
import NebulaIcon from "../../icon/NebulaIcon.vue";

describe("Button", () => {
  it("renders correctly", () => {
    const buttonWrapper = shallowMount(NebulaButton, {
      slots: {
        default: "Follow"
      }
    });
    expect(buttonWrapper.html()).toMatchSnapshot();
  });

  it("renders Chinese characters correctly", () => {
    const iconWrapper = mount(NebulaIcon, {
      attrs: {
        type: "search"
      }
    });

    const buttonWrapper1 = mount(NebulaButton, {
      slots: {
        default: "按钮"
      }
    });
    expect(buttonWrapper1.html()).toMatchSnapshot();

    const buttonWrapper2 = mount(NebulaButton, {
      slots: {
        icon: "search",
        default: "按钮"
      }
    });
    expect(buttonWrapper2.html()).toMatchSnapshot();

    const buttonWrapper3 = mount(NebulaButton, {
      slots: {
        default: iconWrapper.html() + "按钮"
      }
    });
    expect(buttonWrapper3.html()).toMatchSnapshot();

    const buttonWrapper4 = mount(NebulaButton, {
      slots: {
        default: iconWrapper.html() + "按钮"
      },
      propsData: {
        loading: true
      }
    });
    expect(buttonWrapper4.html()).toMatchSnapshot();

    const buttonWrapper5 = mount(NebulaButton, {
      slots: {
        default: "按钮"
      },
      propsData: {
        loading: true
      }
    });
    expect(buttonWrapper5.html()).toMatchSnapshot();
  });

  it("should support link button", () => {
    const wrapper = mount(NebulaButton, {
      propsData: {
        target: "_blank",
        href: "http://nebula.em2046.com/"
      },
      slots: {
        default: "Link button"
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
