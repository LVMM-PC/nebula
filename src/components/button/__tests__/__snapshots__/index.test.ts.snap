// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders Chinese characters correctly 1`] = `<button type="button" class="nebula-btn nebula-btn-two-chinese-chars"><span>按 钮</span></button>`;

exports[`Button renders Chinese characters correctly 2`] = `<button type="button" class="nebula-btn"><span>search按钮</span></button>`;

exports[`Button renders Chinese characters correctly 3`] = `<button type="button" class="nebula-btn"><span><span class="nebula-icon nebula-icon-search"></span>按钮</span></button>`;

exports[`Button renders Chinese characters correctly 4`] = `<button type="button" class="nebula-btn nebula-btn-loading"><span class="nebula-icon nebula-icon-loading nebula-icon-spin"></span><span><span class="nebula-icon nebula-icon-search"></span>按钮</span></button>`;

exports[`Button renders Chinese characters correctly 5`] = `<button type="button" class="nebula-btn nebula-btn-loading"><span class="nebula-icon nebula-icon-loading nebula-icon-spin"></span><span>按钮</span></button>`;

exports[`Button renders correctly 1`] = `<button type="button" class="nebula-btn"><span>Follow</span></button>`;

exports[`Button should support link button 1`] = `<a target="_blank" href="http://nebula.em2046.com/" type="button" class="nebula-btn"><span>Link button</span></a>`;
